### DISCUSSION TITLE
About the data storage period of prometheus

### DISCUSSION THREAD

**Discussion by kousui-dev** [prometheus/prometheus#10326] [Q&A]
We have set 15 days as the data storage period for prometheus.
However, in the graph of grafana, which uses prometheus as a data source
However, in the graph of grafana, which is the data source for prometheus, we can see the data from around 15:00 16 days ago.
You can see not only the data for 15 days (360 hours), which is the data storage period, but also the data from
What could be the reason why the graph in grafana shows the data from around 15:00 16 days ago?
What could be the reason?
Also, I would like to clarify how the retension works to cause this behavior.

prometheus --version
2.22.2

**Comment by roidelapluie** [comment]
The retention period by Prometheus is the minimum retention period. Prometheus stores data in 2 hours blocks that can compacted into larger blocks containing data spanning up to 10% of the retention time, or 31 days, whichever is smaller.

For 15 days, in theory, you could have blocks that could last up to 36 hours.

However, by default, compaction will only create blocks of 2h, 6h, 18h, 54h, 162h and 486h. Therefore, in your setup, blocks will last at most 18h.

It means that you can in theory have data as old as 15 days + 18 hours.

**Comment by kousui-dev** [comment]
Thank you, I understand the behavior of compaction.

I have an additional question about the retention behavior.
In --storage.tsdb.retention.time, the retention period of metrics data can be set not only in days, but also in minutes and hours, but what about the timing to delete old data?

**Reply by roidelapluie** [reply]
We try to delete data every minute if I remember correctly, but we only delete entire blocks. As long as the last timestamp in a block is < 15 days, we keep the block, even if the first timestamp is 15 days + 18 hours.

**Comment by kousui-dev** [comment]
Thank you again for your time.

Is the timestamp of the block based on the minTime and maxTime of meta.json?
Or is it the creation date and time of the block directory?

Is there a way to change the minTime and maxTime of the following data to the exact date and time?

> /prometheus $ cat 01FVAB5TDBQTAFFPN5WEWFD32P/meta.json 
> {
>         "ulid": "01FVAB5TDBQTAFFPN5WEWFD32P",
>         "minTime": 1644170400000,
>         "maxTime": 1644235200000,
>         "stats": {
>                 "numSamples": 760320,
>                 "numSeries": 352,
>                 "numChunks": 6336

I used the date command to change them to UNIX-TIME, Is this output correct?


> $ date --date '@1644170400000'
> Tue Sep  8 09:00:00 JST 54071
> $ date --date '@1644235200000'
> Wed Sep 27 09:00:00 JST 54073

**Reply by kousui-dev** [reply]
I confirmed that the timestamp of the block is based on the minTime and maxTime in meta.json.
I was able to change the minTime and maxTime to UNIX-TIME with the date command. Thank you very much.
The correct result was as follows.


> $ date --date '@1644170400'
> Mon Feb  7 03:00:00 JST 2022
> $ date --date '@1644235200'
> Mon Feb  7 21:00:00 JST 2022
